socklog.8
.TH socklog 8
.SH NAME
socklog \- minimal syslogd replacement for use with djb's daemontools
.SH SYNOPSIS
.B socklog
[
unix
] [
.I path
]
.br
.B socklog
inet [
.I ip
] [
.I port
]
.br
.B socklog
ucspi [
.I args
]
.SH DESCRIPTION
.B socklog
is run under daemontools'
.BR supervise (8),
writing syslog messages it receives from unix domain socket
.I path
or an inet udp socket
.IR ip\fR:\fIport
through a pipe provided by
.BR svscan (8)
to a
.BR multilog (8)
process.
.LP
.B socklog
can be run as an ucspi application to listen to an unix domain stream socket
and for more flexible distributed logging.
.LP
If the environment variables $UID and/or $GID are present,
.B socklog
will drop permissions to those ids after creating and binding the socket
(not in
.I ucspi
mode).
.LP
.B socklog
converts syslog facility and priority information to names
(\fIfacility\fR.\fIpriority\fR:)
as found in
.I /usr/include/syslog.h
at compile time if present.
.SH UNIX SOCKET
.B socklog
[
unix
] [
.I path
]
.LP
Starting
.B socklog
with the 1st argument
.IR unix ,
.B socklog
will listen to the unix domain socket
.IR path .
If
.I path
is omitted, the default
.I /dev/log
is used.
.LP
The 1st argument may be omitted, default is
.IR unix .
.SH INET SOCKET
.B socklog
inet [
.I ip
] [
.I port
]
.LP
Starting
.B socklog
with the 1st argument
.IR inet ,
.B socklog
will listen to the inet udp socket
.IR ip\fR:\fIport .
.LP
If
.I ip
starts with 0,
.B socklog
will bind to all local interfaces.
.LP
If
.I port
is less or equal 1024,
.B socklog
must be run by root.
.LP
.I port
may be omitted, default is 514.
.I ip
may be omitted, default is 0.
.LP
.B socklog
prepends
.I a.b.c.d:
to each syslog message it receives, where
.I a.b.c.d
is the ip address of the connecting system.
.SH UCSPI MODE
.B socklog
ucspi [
.I args
]
.LP
Starting
.B socklog
with the 1st argument
.IR ucspi ,
.B socklog
will run as an ucspi application. Normally
.B socklog
will only be started in
.I ucspi
mode by an ucspi server tool, such as
.BR tcpserver (1)
or
.BR unixserver .
.LP
For each
.IR arg ,
.B socklog
will prepend
$\fIarg\fR:
to each syslog message, if the environment variable
.RI $ arg
is present (maximum is 8).


.SH EXAMPLES
.I unix
mode for listening on unix domain dgram socket
.IR /dev/log .
a proper
.I run
file for
.BR supervise (8)
is:
.LP
  #!/bin/sh
  exec 2>&1
  exec envuidgid nobody socklog unix /dev/log
.LP
.I ucspi
mode for listening on unix domain stream socket
.IR /dev/log .
a proper
.I run
file for
.BR supervise (8)
is:
.LP
  #!/bin/sh
  exec 2>&1
  exec unixserver -U `id -u nobody` /dev/log \\
    socklog ucspi UNIXREMOTEEUID UNIXREMOTEEGID
.LP
.I inet
mode for listening on udp network socket
.IR 0.0.0.0\fR:\fI514 .
a proper
.I run
file for
.BR supervise (8)
is:
.LP
  #!/bin/sh
  exec 2>&1
  exec envuidgid nobody socklog inet 0 514
.LP
.SH AUTHOR
G.Pape <pape@innominate.com>
.SH SEE ALSO
svscan(8),
supervise(8),
multilog(8),
tcpserver(1)

 http://innominate.org/~pape/socklog/
 http://cr.yp.to/daemontools.html
 http://cr.yp.tp/proto/ucspi.txt


package/CHANGES
  * doc/*.html: add color-scheme.
  * doc/benefits.html, doc/notify.html: use chpst -u instead of setuidgid.
  * doc/configuration.html, doc/notify.html, doc/upgrade.html,
    man/socklog-conf.8: use /service instead of /var/service.
  * doc/index.html: spelling.
  * doc/install.html: how to install as non-root user.
  * doc/examples.html, doc/notify.html: upper and lower case, format.
  * doc/*.html, man/*.?: replace; converted from markdown source.
  * socklog-check.c, socklog.c: check length of unix socket to match struct
    sockaddr.
  * pathexec_env.c, socklog-conf.c, socklog.c, tryto.c, uncat.c: whitespace.
  * socklog.c: fix compilation warning -Wpointer-sign.
  * socklog.c, man/socklog.8: add -T to drop timestamps from syslog messages
    (github/leahneukirchen).
  * buffer_put.c, byte.h, byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c,
    byte_rchr.c, byte_zero.c, str_chr.c, str_len.c: sync daemontools library
    fork with runit source.

socklog 2.1.2
Sun, 02 Nov 2025 11:41:11 +0100
  * alloc.c, alloc_re.c, buffer_0.c, byte_chr.c, byte_copy.c, byte_cr.c,
    byte_diff.c, byte_rchr.c, byte_zero.c, error.h, select.h2, wait.h,
    wait_nohang.c, wait_pid.c: sync daemontools library fork with runit
    source.
  * doc/*.html: !DOCTYPE.
  * socklog.c, tryto.c, uncat.c: use parameter name "unused" in sighandler
    functions.
  * hasflock.h1: add #include <unistd.h>.
  * alloc.h, buffer.c, buffer.h, buffer_get.c, buffer_put.c, buffer_write.c,
    byte.h, check-socklog-unix.c, pathexec_env.c, select.h1, select.h2,
    sig.c, sig.h, sig_catch.c, socklog-check.c, socklog-check.dist,
    socklog.c, tryto.c, tryto.dist, uncat.c, uncat.dist, wait.h: fix build
    with gcc15: in C23 f() means f(void), consequently signal handler
    functions now are f(int), alloc_free() and byte_zero() parameters need
    casts, buffer_unixwrite() removes "const" from parameter.
  * package/README: update copyright notice.
  * doc/configuration.html: replace links to daemontools faq with runit faq.
  * doc/*: switch all cr.yp.to http links to https.
  * doc/*, man/*, package/README: switch all smarden.org http links to https.
  * tryto.c: don't report failure if the final try of prog succeeds (thx
    Andras Korn).
  * doc/index.html: remove reference to mailing list misc@list.smarden.org.
  * doc/index.html: add reference to repository on github.com.

socklog 2.1.1
Fri, 11 Oct 2024 00:00:17 +0000
  * doc/install.html: add link to sha256sum.asc.
  * doc/install.html: change mail address.
  * doc/index.html: remove reference to socklog 2.0.x.
  * doc/*.html: remove <small>$Id$</small>.
  * debian/: remove; obsolete.
  * pathexec.h, pathexec_env.c, pathexec_run.c, seek_set.c, sgetopt.c,
    sgetopt.h, socklog-check.c, socklog-conf.c, socklog.c, subgetopt.c,
    subgetopt.h, trycpp.c, tryflock.c, trypoll.c, trysgact.c, trysgprm.c,
    tryto.c, tryulong64.c, trywaitp.c, uncat.c, x86cpuid.c: incorporate
    lib changes from runit-2.2.0 and adapt socklog programs to properly
    build with modern toolchains (thx Gentoo, Debian, Void Linux).
  * prot.c, prot.h, chkshsgr.c, tryshsgr.c, hasshsgr.h?, warn-shsgr:
    remove; obsolete.
  * Makefile, TARGETS: no longer check for "shortsetgroups" (thx Leah
    Neukirchen).
  * socklog.c: replace prot_gid() with setgroups() and prot_uid() with
    setuid().

socklog 2.1.0
Mon, 06 Mar 2006 12:59:33 +0000
  * socklog-conf, man/socklog-conf.8: create service directories
    /etc/sv/socklog-* instead of /etc/socklog/*; create ./check script
    for socklog-unix service.
  * doc/configuration.html, doc/install.html, doc/network.html,
    doc/notify.html: adapt to new path to service directories; utilize
    runit's sv program instead of runsvctrl.
  * man/socklog-check.8, man/socklog-conf.8, man/socklog.8, man/uncat.1:
    see also sv(8).

socklog 2.0.3
Sun, 26 Feb 2006 13:29:17 +0000
  * doc/index.html: minor.
  * socklog-check.c: change EDESTADDRREQ to ECONNREFUSED if defined and set
    by connect().
  * tryto: fix a race with child terminating immediatly (thx Alex Efros);
    catch SIGCHLD while waiting ktimeout seconds before sending KILL.
  * man/tryto.1: remove reference to timelimit(8) (thx Alex Efros).
  * socklog.check, socklog.dist: don't check usage output with -V as it
    differs on solaris (thx Asif Iqbal).

socklog 2.0.2
Sun, 10 Apr 2005 07:21:30 +0000
  * doc/readme.solaris.html: add -ldoor -lthread to src/conf-ld (thx Lars
    Kellogg-Stedman); update log/run script, mailing list address.
  * man/socklog.8: remove double entries in SEE ALSO.

socklog 2.0.1
Thu, 10 Feb 2005 10:05:55 +0000
  * Makefile: add linker options detected by trysocketlib.c to targets
    check-socklog-inet, check-socklog-inet (fixes test failure on solaris,
    thx Uffe Jakobsen).

socklog 2.0.0
Sun, 06 Feb 2005 19:57:04 +0000
  * socklog-check.check: tweak timing in program test.
  * tryto.c: new option -P: run prog in new session, and signal process
    group.
  * man/tryto.1: adapt; cleanup.
  * man/socklog-check.8: add hint about runit services using syslog
    facility; cleanup.
  * man/socklog-conf.8, man/socklog.8, man/uncat.1: cleanup.
  * doc/index.html: new stable release.
  * doc/network.html: typo.

socklog 1.5.0
Sat, 26 Jun 2004 11:28:33 +0000
  * socklog-check, man/socklog-check.8: new; check for the availability of
    a socklog service.
  * socklog-check.check, socklog-check.dist: new; program test.
  * byte_zero.c: new.
  * socklog.c, man/socklog.8: new option -U: don't override umask when
    creating domain socket.
  * package/check, package/compile, package/install-man, package/upgrade:
    minor cleanup.
  * Makefile, man/socklog-conf.8: adapt.

socklog 1.4.2
Tue, 23 Mar 2004 18:16:17 +0000
  * doc/index.html, doc/install.html: minor.
  * socklog.c: fix sigterm handling in ucspi mode.
  * uncat.check: tweak timing in program test.
  * Makefile: don't run target check by default.
  * package/check: new; run checks on programs.

src/socklog.c
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/un.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>

#include "buffer.h"
#include "error.h"
#include "strerr.h"
#include "scan.h"
#include "env.h"
#include "prot.h"
#include "sig.h"
#include "open.h"
#include "sgetopt.h"

#define SYSLOG_NAMES
#include <syslog.h>

#if defined(__sun__) && defined(__sparc__) && defined(__unix__) && defined(__svr4__)
#define SOLARIS
# include <stropts.h>
# include <sys/strlog.h>
# include <fcntl.h>
# include "syslognames.h"
#if WANT_SUN_DOOR
# include <door.h>
#endif
#endif

/* #define WARNING "socklog: warning: " */
#define FATAL "socklog: fatal: "

#ifdef SOLARIS
#define USAGE " [-rR] [unix|inet|ucspi|sun_stream] [args]"
#else
#define USAGE " [-rR] [unix|inet|ucspi] [args]"
#endif

#define VERSION "$Id$"
#define DEFAULTINET "0"
#define DEFAULTPORT "514"
#define DEFAULTUNIX "/dev/log"

const char *progname;

#define LINEC 1024
#define MODE_UNIX 0
#define MODE_INET 1
#define MODE_UCSPI 2
#ifdef SOLARIS
#define MODE_SUN_STREAM 3
#endif

int mode =MODE_UNIX;
char line[LINEC];
const char *address =NULL;
char *uid, *gid;
unsigned int lograw =0;

int flag_exitasap = 0;
void sig_term_catch(void) {
  flag_exitasap = 1;
}

void usage() {
  strerr_die4x(1, "usage: ", progname, USAGE, "\n");
}

void out(const char *s1, const char *s2) {
  if (s1) buffer_puts(buffer_1, s1);
  if (s2) buffer_puts(buffer_1, s2);
}
void err(const char *s1, const char *s2, const char *s3) {
  if (s1) buffer_puts(buffer_2, s1);
  if (s2) buffer_puts(buffer_2, s2);
  if (s3) buffer_puts(buffer_2, s3);
}

void setuidgid() {
  /* drop permissions */
  if ((gid = env_get("GID")) != NULL) {
    unsigned long g;

    scan_ulong(gid, &g);
    err("gid=", gid, ", ");
    if (prot_gid(g) == -1)
      strerr_die2sys(111, FATAL, "unable to setgid: ");
  }
  if ((uid = env_get("UID")) != NULL) {
    unsigned long u;

    scan_ulong(uid, &u);
    err("uid=", uid, ", ");
    if (prot_uid(u) == -1)
      strerr_die2sys(111, FATAL, "unable to setuid: ");
  }
}

int print_syslog_names(int fpr, buffer *buf) {
  int fp =LOG_FAC(fpr) <<3;
  CODE *p;
  int rc =1;
  for (p =facilitynames; p->c_name; p++) {
    if (p->c_val == fp) {
      buffer_puts(buf, p->c_name);
      buffer_puts(buf, ".");
      break;
    }
  }
  if (! p->c_name) {
    buffer_puts(buf, "unknown.");
    rc =0;
  }
  fp =LOG_PRI(fpr);
  for (p =prioritynames; p->c_name; p++) {
    if (p->c_val == fp) {
      buffer_puts(buf, p->c_name);
      buffer_puts(buf, ": ");
      break;
    }
  }
  if (! p->c_name) {
    buffer_puts(buf, "unknown: ");
    rc =0;
  }
  return(rc);
}

int scan_syslog_names (char *l, int lc, buffer *buf) {
  int i;
  int ok =0;
  int fpr =0;

  if (l[0] != '<') return(0);
  for (i =1; (i < 5) && (i < lc); i++) {
    if (l[i] == '>') {
      ok =1;
      break;
    }
    if (('0' <= l[i]) && (l[i] <= '9')) {
      fpr =10 *fpr + l[i] -'0';
    } else {
      return(0);
    }
  }
  if (!ok || !fpr) return(0);
  return(print_syslog_names(fpr, buf) ? ++i : 0);
}

void remote_info (struct sockaddr_in *sa) {
  char *host;

  host =inet_ntoa(sa->sin_addr);
  out(host, ": ");
}

int socket_unix (const char* f) {
  int s;
  struct sockaddr_un sa;
  
  if ((s =socket(AF_UNIX, SOCK_DGRAM, 0)) == -1)
    strerr_die2sys(111, FATAL, "socket(): ");

  sa.sun_family =AF_UNIX;
  strncpy(sa.sun_path, f, sizeof(sa.sun_path));
  unlink(f);
  umask(0);
  if (bind(s, (struct sockaddr*) &sa, sizeof sa) == -1)
    strerr_die2sys(111, FATAL, "bind(): ");

  err("listening on ", f, ", ");
  return(s);
}

int socket_inet (const char* ip, const char* port) {
  int s;
  unsigned long p;
  struct sockaddr_in sa;
  
  if (ip[0] == '0') {
    sa.sin_addr.s_addr =INADDR_ANY;
  } else {
#ifndef SOLARIS
    if (inet_aton(ip, &sa.sin_addr) == 0) {
      strerr_die2sys(111, FATAL, "inet_aton(): ");
    }
#else
    sa.sin_addr.s_addr =inet_addr(ip);
#endif
  }
  if ((s =socket(AF_INET, SOCK_DGRAM, 0)) == -1)
    strerr_die2sys(111, FATAL, "socket(): ");
  if (scan_ulong(port, &p) == 0)
    strerr_die3x(111, FATAL, "bad port number: ", port);

  sa.sin_family =AF_INET;
  sa.sin_port =htons(p);
  if (bind(s, (struct sockaddr*) &sa, sizeof sa) == -1)
    strerr_die2sys(111, FATAL, "bind(): ");

  ip =inet_ntoa(sa.sin_addr);
  err("listening on ", ip, 0);
  err(":", port, ", ");
  return(s);
}

int read_socket (int s) {
  sig_catch(sig_term, sig_term_catch);
  sig_catch(sig_int, sig_term_catch);
  /* drop permissions */
  setuidgid();
  buffer_putsflush(buffer_2, "starting.\n");

  for(;;) {
    struct sockaddr_in saf;
    int dummy =sizeof saf;
    int linec;
    int os;
    
    linec =recvfrom(s, line, LINEC, 0, (struct sockaddr *) &saf, &dummy);
    if (linec == -1) {
      if (errno != error_intr)
	strerr_die2sys(111, FATAL, "recvfrom(): ");
      else
	linec =0;
    }
    if (flag_exitasap) break;

    while (linec && (line[linec -1] == 0)) linec--;
    if (linec == 0) continue;

    if (lograw) {
      buffer_put(buffer_1, line, linec);
      if (line[linec -1] != '\n') {
	if (linec == LINEC) out("...", 0);
	out("\n", 0);
      }
      if (lograw == 2) {
	buffer_flush(buffer_1);
	continue;
      }
    }

    if (mode == MODE_INET) remote_info(&saf);
    os =scan_syslog_names(line, linec, buffer_1);

    buffer_put(buffer_1, line +os, linec -os);
    if (line[linec -1] != '\n') {
      if (linec == LINEC) out("...", 0);
      out("\n", 0);
    }
    buffer_flush(buffer_1);
  }
  return(0);
}

int read_ucspi (int fd, const char **vars) {
  char *envs[9];
  int flageol =1;
  int i;
  
  for (i =0; *vars && (i < 8); vars++) {
    if ((envs[i] =env_get(*vars)) != NULL)
      i++;
  }
  envs[i] =NULL;
  
  for(;;) {
    int linec;
    char *l, *p;
    
    linec =buffer_get(buffer_0, line, LINEC);
    if (linec == -1)
      strerr_die2sys(111, FATAL, "read(): ");

    if (linec == 0) {
      if (! flageol) err("\n", 0, 0);
      buffer_flush(buffer_2);
      return(0);
    }
    
    for (l =p =line; l -line < linec; l++) {
      if (flageol) {
	if (! *l || (*l == '\n')) continue;
	for (i =0; envs[i]; i++) {
	  err(envs[i], ": ", 0);
	}
	/* could fail on eg <13\0>user.notice: ... */
	l += scan_syslog_names(l, line -l +linec, buffer_2);
	p =l;
	flageol =0;
      }
      if (! *l || (*l == '\n')) {
	buffer_put(buffer_2, p, l -p);
	buffer_putsflush(buffer_2, "\n");
	flageol =1;
      }
    }
    if (!flageol) buffer_putflush(buffer_2, p, l -p);
  }
}

#ifdef SOLARIS
#if WANT_SUN_DOOR
static void door_proc(void *cookie, char *argp, size_t arg_size,
		      door_desc_t *dp, uint_t ndesc) {
  door_return(NULL, 0, NULL, 0);
  return;
}

static int door_setup(const char *door) {
  int dfd;
  struct door_info di;

  if ( (dfd = open_trunc(door)) == -1)
    strerr_die2sys(111, FATAL, "open_trunc(): ");

  if (door_info(dfd, &di) == -1) {
    if (errno != EBADF)
      strerr_die2sys(111, FATAL, "door_info(): ");
  }
  else {
    /*XXX: could log the pid of the door owner. */
    if (di.di_target != -1)
      strerr_die4x(100, FATAL, "door ", door, " allready in use.");
  }
  
  close(dfd);
  fdetach(door); /* highjack the door file */

  if ((dfd =door_create(door_proc, NULL, 0)) == -1)
    strerr_die2sys(111, FATAL, "door_create(): ");

  if (fattach(dfd, door) == -1)
    strerr_die2sys(111, FATAL, "fattach(): ");

  err("door path is ", door, ", ");
  return(dfd);
}
#endif /*WANT_SUN_DOOR*/

static int stream_sun(const char *address) {
  int sfd;
  struct strioctl sc;

  if ((sfd = open(address, O_RDONLY | O_NOCTTY)) == -1)
    strerr_die2sys(111, FATAL, "open(): ");
  
  memset(&sc, 0, sizeof(sc));
  sc.ic_cmd =I_CONSLOG;
  if (ioctl(sfd, I_STR, &sc) < 0)
    strerr_die2sys(111, FATAL, "ioctl(): ");

  err("sun_stream is ", address, ", ");
  return(sfd);
}

static void read_stream_sun(int fd) {
  struct strbuf ctl, data;
  struct log_ctl logctl;
  int flags;
  
  ctl.maxlen =ctl.len =sizeof(logctl);
  ctl.buf =(char *) &logctl;
  data.maxlen =LINEC;
  data.len =0;
  data.buf =line;
  flags =0;
  
  sig_catch(sig_term, sig_term_catch);
  sig_catch(sig_int, sig_term_catch);
  setuidgid();
  buffer_putsflush(buffer_2, "starting.\n");
  
  /* read the messages */
  for (;;) {

    if ((getmsg(fd, &ctl, &data, &flags) & MORECTL) && (errno != error_intr))
      strerr_die2sys(111, FATAL, "getmsg(): ");

    if (flag_exitasap)
      return;

    if (data.len) {
      int shorten =data.len;
      if (!line[shorten-1])
        shorten--;
      while (line[shorten-1] == '\n')
        shorten--;

      (void) print_syslog_names(logctl.pri, buffer_1);

      buffer_put(buffer_1, line, shorten);
      if (data.len == LINEC) out("...", 0);
      out("\n", 0);

      buffer_flush(buffer_1);
    }
  }
}

#endif

int main(int argc, const char **argv, const char *const *envp) {
  int opt;
  int s =0;
  
  progname =*argv;

  while ((opt =getopt(argc, argv, "rRV")) != opteof) {
    switch(opt) {
    case 'r': lograw =1; break;
    case 'R': lograw =2; break;
    case 'V':
      err(VERSION, 0, 0);
      buffer_putsflush(buffer_2, "\n\n");
    case '?': usage();
    }
  }
  argv +=optind;

  if (*argv) {
    switch (**argv) {
    case 'u':
      if (! *(++*argv)) usage();
      switch (**argv) {
      case 'n':
	mode =MODE_UNIX;
	break;
      case 'c':
	mode =MODE_UCSPI;
	argv--;
	break;
      default:
	usage();
      }
      break;
    case 'i':
      mode =MODE_INET;
      break;
#ifdef SOLARIS
    case 's':
      mode =MODE_SUN_STREAM;
      break;
#endif
    default:
      usage();
    }
    argv++;
  }

  if (*argv) address =*argv++;

  switch (mode) {
  case MODE_INET: {
    const char* port =NULL;

    if (*argv) port =*argv++;
    if (*argv) usage();
    if (!address) address =DEFAULTINET;
    if (!port) port =DEFAULTPORT;
    s =socket_inet(address, port);
    return(read_socket(s));
  }
  case MODE_UNIX: {
    if (*argv) usage();
    if (!address) address =DEFAULTUNIX;
    s =socket_unix(address);
    return(read_socket(s));
  }
#ifdef SOLARIS
  case MODE_SUN_STREAM: {
#if WANT_SUN_DOOR
    const char *door =NULL;
    int dfd =-1;
    if (*argv) door =*argv++;
#endif
    if (!address) address =DEFAULTUNIX;
    if (*argv) usage();

    s =stream_sun(address);

#if WANT_SUN_DOOR
    if (door)
      dfd = door_setup(door);
#endif

    read_stream_sun(s);

#if WANT_SUN_DOOR
    if (dfd != -1)
      door_revoke(dfd);
    /*
    ** syslogd does unlink() the door file, but we can't, since we droped
    ** all privs before.
    */
#endif
    return 0;
  }
#endif /*SOLARIS*/
  case MODE_UCSPI:
    s =0;
    return(read_ucspi(0, argv));
  }
  /* not reached */
  return(1);
}
